### ISSUE TITLE
dune crash when deleting temp files

### ISSUE THREAD

**Issue by avsm** [ocaml/dune#4489]
Spotted in an opam-repo test in https://github.com/ocaml/opam-repository/pull/18483, full log at https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/766a299dae4c604d7fd0e72a3236a01dd16a65ed/variant/compilers,4.11,extunix.0.3.1,tests

```
- Internal error, please report upstream including the contents of _build/log.
- Description:
-   ("hooks failed",
-   { exns =
-       [ { exn =
-             "Unix.Unix_error(Unix.ENOENT, \"unlink\", \"/tmp/build2b91e7.dune/extunixac1694write\")"
-         ; backtrace =
-             "Raised by primitive operation at Stdune__path.clear_dir.(fun) in file \"src/stdune/path.ml\", line 1163, characters 15-34\n\
-              Called from Stdlib__list.iter in file \"list.ml\", line 110, characters 12-15\n\
-              Called from Stdune__path.clear_dir in file \"src/stdune/path.ml\" (inlined), line 1178, characters 20-45\n\
-              Called from Stdune__temp.clear_dir in file \"src/stdune/temp.ml\", line 84, characters 2-20\n\
-              Called from Stdune__exn_with_backtrace.try_with in file \"src/stdune/exn_with_backtrace.ml\", line 9, characters 8-12\n\
-              "
-         }
-       ]
-   })
- Raised at Stdune__code_error.raise in file "src/stdune/code_error.ml", line
-   9, characters 30-62
- Called from Stdlib.at_exit.(fun) in file "stdlib.ml", line 553, characters
-   62-65
- Called from Stdlib.do_at_exit in file "stdlib.ml" (inlined), line 556,
-   characters 20-39
- Called from Stdlib.exit in file "stdlib.ml", line 559, characters 2-15
- Called from Main in file "bin/main.ml", line 264, characters 11-17
- 
- I must not crash.  Uncertainty is the mind-killer. Exceptions are the
- little-death that brings total obliteration.  I will fully express my cases. 
- Execution will pass over me and through me.  And when it has gone past, I
- will unwind the stack along its path.  Where the cases are handled there will
- be nothing.  Only I will remain.
```

**Comment by jeremiedimino** [comment]
This might be fixed in the tip of the repo as we did some work on temporary files.

**Comment by rgrinberg** [comment]
I'm not so certain that is fixed, but I wonder if we perhaps ignore ENOENT errors when clearing temp files? That seems safe enough.

**Comment by jeremiedimino** [comment]
Agreed

**Comment by snowleopard** [comment]
I came across a similar crash and fixed it in #4529, so I think we can close this now.

There was a race in `clear_dir`: a concurrent process could delete an entry after we saw it in the listing of a directory in `clear_dir` but before we managed to delete it ourselves.
